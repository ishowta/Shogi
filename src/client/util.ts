export const isSameInstance = <T>(a: T, b: T): boolean => Object.is(a, b)
