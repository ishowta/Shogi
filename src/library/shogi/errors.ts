/** 将棋ライブラリ内エラー */
export class ShogiError implements Error {

    /** エラーメッセージ */
    public message: string
    /** エラー名 */
    public name: string = "ShogiError"

    public constructor(message: string = "") {
        this.message = message
    }

    /** エラー名とメッセージを文字列にして返す */
    public toString(): string {
        return `${this.name} : ${this.message}`
    }
}

/** 指した時のエラー（物理エラー） */
export class MoveError extends ShogiError { }

/** 範囲から出ている */
export class BoundError extends MoveError { }
/** 重複しているところに置こうとした */
export class DuplicateError extends MoveError { }
/** 相手の駒を動かそうとした */
export class NotOwnedPieceError extends MoveError { }
/** 存在しない駒を動かそうとした */
export class NoPieceError extends MoveError { }
/** 動かせない場所に動かそうとした */
export class CantMoveError extends MoveError { }
/** 成れないのに成ろうとした */
export class CantPromoteError extends MoveError { }

/** 指した時の反則（論理エラー） */
export class FoulError extends ShogiError { }

/** 千日手 */
export class ThousandDaysFoul extends FoulError { }
/** 王手放置 */
export class NeglectKingFoul extends FoulError { }
/** 打ち歩詰め */
export class StrikingFoul extends FoulError { }
/** ２歩 */
export class DoublePawnFoul extends FoulError { }
