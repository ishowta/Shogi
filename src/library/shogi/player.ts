/**
 * プレイヤーの区別
 *
 * 先行は黒番
 */
export enum Player {
    Black = "black",
    White = "white",
}
