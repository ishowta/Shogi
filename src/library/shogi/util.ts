/* tslint:disable:no-any no-unsafe-any binary-expression-operand-order one-variable-per-declaration typedef no-for-in */
/**
 * オブジェクトをdeepCopyする
 *
 * 参考：https://stackoverflow.com/questions/28150967/typescript-cloning-object
 * @param obj オブジェクト
 */
export const deepCopy = <T>(obj: T): T => {
    let copy: any

    // Handle the 3 simple types, and null or undefined
    if (null === obj || "object" !== typeof obj) { return obj }

    // Handle Date
    if (obj instanceof Date) {
        copy = new Date()
        copy.setTime(obj.getTime())
        copy.__proto__ = (obj as any).__proto__
        return copy
    }

    // Handle Array
    if (obj instanceof Array) {
        copy = []
        for (let i = 0, len = obj.length; i < len; i += 1) {
            copy[i] = deepCopy(obj[i])
        }
        copy.__proto__ = (obj as any).__proto__
        return copy
    }

    // Handle Object
    if (obj instanceof Object) {
        copy = {}
        for (const attr in obj) {
            if ((obj as any).hasOwnProperty(attr)) { copy[attr] = deepCopy(obj[attr]) }
        }
        copy.__proto__ = (obj as any).__proto__
        return copy
    }

    throw new Error("Unable to copy obj! Its type isn't supported.")
}

/**
 * 範囲を指定して配列を生成する
 *
 * @param start 左端
 * @param end 右端
 */
export const range = (start: integer, end: integer): integer[] =>
    Array.from({ length: (end - start + 1) }, (_, k) => k + start)

/**
 * 範囲を指定してイテレーターを生成する
 *
 * @param start 左端
 * @param end 右端
 */
export function* irange(start: integer, end: integer): IterableIterator<integer> {
    for (let i = start; i <= end; i += 1) {
        yield i
    }
}

/** 小さい方の値を返す */
export const min = (a: integer, b: integer) => a <= b ? a : b

/** 大きい方の値を返す */
export const max = (a: integer, b: integer) => a > b ? a : b

/** ２次元座標 */
export type Point = {
    /** x軸 */
    readonly x: integer
    /** y軸 */
    readonly y: integer
}

export const add = (a: Point, b: Point) => ({x: a.x + b.x, y: a.y + b.y})
export const sub = (a: Point, b: Point) => ({x: a.x - b.x, y: a.y - b.y})
export const mul = (a: Point, b: Point) => ({x: a.x * b.x, y: a.y * b.y})
export const div = (a: Point, b: Point) => ({x: a.x / b.x, y: a.y / b.y})

/** aとbの間にある座標のリストを返す（aやbは含まない） */
export const interpolation = (c: integer, d: integer): integer[] =>
        range(min(c, d) + 1, max(c, d) - 1)

/** aとbの間にある座標のリストを返す（８方向のみで、aやbは含まない） */
export const interpolation2D = (a: Point, b: Point): Point[] => {
    const rangeX: integer[] = a.x < b.x ? interpolation(a.x, b.x) : interpolation(a.x, b.x).reverse()
    const rangeY: integer[] = a.y < b.y ? interpolation(a.y, b.y) : interpolation(a.y, b.y).reverse()
    return rangeX.map<Point>((_, i) => ({x: rangeX[i], y: rangeY[i]}))
}

/** オブジェクトの要素が同一であるか */
export const equal = (a: object, b: object): boolean => JSON.stringify(a) === JSON.stringify(b)

/** Syntax sugar of a === b */
export const isSameInstance = <T>(a: T, b: T): boolean => Object.is(a, b)
