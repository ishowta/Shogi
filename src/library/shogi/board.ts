import { Piece, PieceType } from "./piece"
import { Player } from "./player"
import { deepCopy, Point, range, irange } from "./util"

/** 将棋盤 */
export class Board<T> extends Array<T[]> {
	/** 高さ（上下の長さ） */
	public static readonly height: integer = 9

	/** 幅（左右の長さ） */
	public static readonly width: integer = 9

	/** 2次元配列で初期化 */
	protected resetByArray(arr: T[][]): void {
		this.length = 0
		for (const y of range(0, Board.height - 1)) {
			this.push(new Array<T>())
			const line: T[] = this[this.length - 1]
			for (const x of range(0, Board.width - 1)) {
				line.push(arr[y][x])
			}
		}
	}

	/** defaultValueで盤面を初期化 */
	protected resetByValue(defaultValue: T): void {
		for (const y of range(0, Board.height - 1)) {
			this.push(new Array<T>())
			const line: T[] = this[this.length - 1]
			for (const x of range(0, Board.width - 1)) {
				line.push(defaultValue)
			}
		}
	}

	/** 座標を値として入れた盤面を作成 */
	public static posMatrix(): Board<Point> {
		const posBoard: Board<Point> = new Board<Point>()
		posBoard.resetByArray(range(0, Board.height - 1).map(y => range(0, Board.width - 1).map(x => ({ x, y }))))
		return posBoard
	}

	/**
	 * 位置pの要素を返す
	 * @param p 位置
	 */
	public at(p: Point): T | null {
		if (!Board.inBound(p)) {
			return null
		}
		return this[p.y][p.x]
	}

	/**
	 * 位置pにeを代入する
	 * @param p 代入する位置
	 * @param e 対象
	 */
	public assign(p: Point, e: T): this {
		this[p.y][p.x] = e
		return this
	}

	/**
	 * 位置pが盤面の中を指しているかチェック
	 * @param p 位置
	 */
	public static inBound(p: Point): boolean {
		return !(p.x < 0 || p.x >= Board.width || p.y < 0 || p.y >= Board.height)
	}

	/**
	 * 盤面をひっくり返して関数を実行する
	 * @param fn 逆向きで実行する関数
	 */
	public applyWithReverse<U>(fn: () => U): U {
		this.reverseMatrix()
		const res: U = fn()
		this.reverseMatrix()
		return res
	}

	/** matrix map */
	public matMap<U>(fn: (e: T, index: Point) => U): U[][] {
		return this.map((line, y) => line.map((e, x) => fn(e, { x, y })))
	}

	/** matrix reduce */
	public matReduce<U>(
		fn: (previousValue: U, currentValue: T, currentIndex: Point, array: T[]) => U,
		initialValue: U
	): U {
		return this.reduce((p, c, y) => c.reduce((pp, cc, x, arr) => fn(pp, cc, { x, y }, arr), p), initialValue)
	}

	/** matrix some */
	public matSome(fn: (e: T, index: Point) => boolean): boolean {
		return this.some((line, y) => line.some((e, x) => fn(e, { x, y })))
	}

	/** matrix every */
	public matEvery(fn: (e: T, index: Point) => boolean): boolean {
		return this.every((line, y) => line.every((e, x) => fn(e, { x, y })))
	}

	/** matrix filter */
	public matFilter(fn: (e: T, index: Point) => boolean): T[] {
		return this.flatMap((line, y) => line.filter((e, x) => fn(e, { x, y })))
	}

	/** matrix find */
	public matFind(fn: (e: T, index: Point) => boolean): T | null {
		for (const [y, line] of this.entries()) {
			for (const [x, e] of line.entries()) {
				if (fn(e, { x, y })) {
					return e
				}
			}
		}
		return null
	}

	/** matrix findIndex */
	public matFindIndex(fn: (e: T, index: Point) => boolean): Point | null {
		for (const [y, line] of this.entries()) {
			for (const [x, e] of line.entries()) {
				if (fn(e, { x, y })) {
					return { x, y }
				}
			}
		}
		return null
	}

	/** matrix forEach */
	public matForEach(fn: (e: T, index: Point) => void): void {
		this.forEach((line, y) => line.forEach((e, x) => fn(e, { x, y })))
	}

	/** matrix zip (slow) */
	public zip(rhs: Board<T>): Board<[T, T]> {
		const b: Board<[T, T]> = new Board<[T, T]>()
		b.resetByArray(
			this.matMap((e, pos) => {
				const t: [T, T] = [e, rhs.at(pos) as T]
				return t
			})
		)
		return b
	}

	/** 反転させる */
	private reverseMatrix(): void {
		const bufThis: this = deepCopy(this)
		for (const y of irange(0, Board.height - 1)) {
			for (const x of irange(0, Board.width - 1)) {
				this[Board.height - 1 - y][Board.width - 1 - x] = bufThis[y][x]
			}
		}
	}
}

/** 駒の将棋盤 */
export class PieceBoard extends Board<Piece | null> {
	/** 将棋盤を配列で初期化 */
	public static init(arr: Array<Array<Piece | null>>): PieceBoard {
		const pieceBoard: PieceBoard = new PieceBoard()
		pieceBoard.resetByArray(arr)
		return pieceBoard
	}

	/** 将棋盤を初期化 */
	public static make(): PieceBoard {
		const place = (board: PieceBoard, player: Player): void => {
			board[8][0] = new Piece(PieceType.Lance, player)
			board[8][1] = new Piece(PieceType.Knight, player)
			board[8][2] = new Piece(PieceType.SilverGeneral, player)
			board[8][3] = new Piece(PieceType.GoldGeneral, player)
			board[8][4] = new Piece(PieceType.King, player)
			board[8][5] = new Piece(PieceType.GoldGeneral, player)
			board[8][6] = new Piece(PieceType.SilverGeneral, player)
			board[8][7] = new Piece(PieceType.Knight, player)
			board[8][8] = new Piece(PieceType.Lance, player)
			board[7][1] = new Piece(PieceType.Bishop, player)
			board[7][7] = new Piece(PieceType.Rook, player)
			board[6] = board[6].map(_ => new Piece(PieceType.Pawn, player))
		}

		const pieceBoard: PieceBoard = new PieceBoard()
		pieceBoard.resetByValue(null)

		place(pieceBoard, Player.Black)
		pieceBoard.applyWithReverse(() => place(pieceBoard, Player.White))
		return pieceBoard
	}
}

/** Boolean将棋盤 */
export class BitBoard extends Board<boolean> {
	/** 初期化 */
	public static init(arr?: boolean[][]): BitBoard {
		const pieceBoard: BitBoard = new BitBoard()
		if (arr !== undefined) {
			pieceBoard.resetByArray(arr)
		} else {
			pieceBoard.resetByValue(false)
		}
		return pieceBoard
	}

	/** bit & */
	public and(rhs: BitBoard): BitBoard {
		return BitBoard.init(this.zip(rhs).matMap(([a, b]) => a && b))
	}

	/** rhsでbitboardにマスクを掛ける */
	public mask: (rhs: BitBoard) => BitBoard = this.and

	/** bit | */
	public or(rhs: BitBoard): BitBoard {
		return BitBoard.init(this.zip(rhs).matMap(([a, b]) => a || b))
	}

	/** trueになっている場所が存在するか */
	public exist(): boolean {
		return this.matSome(b => b)
	}
}
