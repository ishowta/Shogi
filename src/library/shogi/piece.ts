import { Board } from "./board"
import { Player } from "./player"
import { Point } from "./util"

/** 駒の種類 */
export enum PieceType {
    /** 王将 */
    King =  "king",
    /** 飛車 */
    Rook =  "rook",
    /** 角行 */
    Bishop = "bishop",
    /** 金将 */
    GoldGeneral= "gold-general",
    /** 銀将 */
    SilverGeneral= "silver-general",
    /** 桂馬 */
    Knight= "knight",
    /** 香車 */
    Lance= "lance",
    /** 歩兵 */
    Pawn= "pawn",
}

/** 駒の動ける方向と回数 */
export type Dir = {
    /** 駒の動ける場所 dP/dt */
    dif: Point,

    /** 駒の動ける回数 t */
    times: integer
}

/** 駒 */
export class Piece {
    /** 駒の種類（成っているかは含めない） */
    public readonly type: PieceType

    /** 駒の所持プレイヤー */
    public owner: Player

    /** 成っているか */
    public isPromote: boolean

    public constructor(type: PieceType, owner: Player) {
        this.type = type
        this.owner = owner
        this.isPromote = false
    }

    /** 駒から見た盤面の位置（白番の場合反転する） */
    public look(p: Point): Point {
        return {x: this.lookX(p.x), y: this.lookY(p.y)}
    }

    /** 駒から見た盤面の位置X（白番の場合反転する） */
    public lookX(x: integer): integer {
        return this.owner === Player.Black ? x : (Board.width - 1) - x
    }

    /** 駒から見た盤面の位置Y（白番の場合反転する） */
    public lookY(y: integer): integer {
        return this.owner === Player.Black ? y : (Board.height - 1) - y
    }

    /** 文字に変換 */
    public toString(): string {
        switch (this.type) {
            case PieceType.Bishop:
                return this.isPromote ? "馬" : "角"
            case PieceType.GoldGeneral:
                return this.isPromote ? "？" : "金"
            case PieceType.King:
                return this.isPromote ? "？" : this.owner === Player.Black ? "王" : "玉"
            case PieceType.Knight:
                return this.isPromote ? "金" : "桂"
            case PieceType.Lance:
                return this.isPromote ? "金" : "香"
            case PieceType.Pawn:
                return this.isPromote ? "と" : "歩"
            case PieceType.Rook:
                return this.isPromote ? "竜" : "飛"
            case PieceType.SilverGeneral:
                return this.isPromote ? "金" : "銀"
        }
    }

    /** 成れる種類の駒か */
    public static canPromote(type: PieceType): boolean {
        switch (type) {
            case PieceType.King: return false
            case PieceType.Rook: return true
            case PieceType.Bishop: return true
            case PieceType.GoldGeneral: return false
            case PieceType.SilverGeneral: return true
            case PieceType.Knight: return true
            case PieceType.Lance: return true
            case PieceType.Pawn: return true
        }
    }

    /** 動ける場所と回数の表 */
    private static MovedirMatrix: {[key in PieceType]: [integer[][], integer[][]]} = {
        [PieceType.Pawn] : [[
            [0, 1, 0],
            [0, 0, 0],
            [0, 0, 0],
        ], [
            [1, 1, 1],
            [1, 0, 1],
            [0, 1, 0],
        ], ],
        [PieceType.Lance] : [[
            [0, Infinity, 0],
            [0, 0, 0],
            [0, 0, 0],
        ], [
            [1, 1, 1],
            [1, 0, 1],
            [0, 1, 0],
        ], ],
        [PieceType.Knight] : [[
            [0, 1, 0, 1, 0],
            [0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0],
            [0, 1, 0, 1, 0],
        ], [
            [1, 1, 1],
            [1, 0, 1],
            [0, 1, 0],
        ], ],
        [PieceType.SilverGeneral] : [[
            [1, 1, 1],
            [0, 0, 0],
            [1, 0, 1],
        ], [
            [1, 1, 1],
            [1, 0, 1],
            [0, 1, 0],
        ], ],
        [PieceType.GoldGeneral] : [[
            [1, 1, 1],
            [1, 0, 1],
            [0, 1, 0],
        ], [], ],
        [PieceType.Rook] : [[
            [0       , Infinity, 0       ],
            [Infinity, 0       , Infinity],
            [0       , Infinity, 0       ],
        ], [
            [1       , Infinity, 1       ],
            [Infinity, 0       , Infinity],
            [1       , Infinity, 1       ],
        ], ],
        [PieceType.Bishop] : [[
            [Infinity, 0       , Infinity],
            [0       , 0       , 0       ],
            [Infinity, 0       , Infinity],
        ], [
            [Infinity, 1       , Infinity],
            [1       , 0       , 1       ],
            [Infinity, 1       , Infinity],
        ], ],
        [PieceType.King] : [[
            [1, 1, 1],
            [1, 0, 1],
            [1, 1, 1],
        ], [], ],
    }

    /**
     * 駒の動かせる方向と回数のリスト
     *
     * 数値は何回動かせるかを示す
     */
    public static MovableDir: {[pieceTypeKey in PieceType]: {[isPromotedKey in  "normal" | "promoted"]: Dir[]}} =
    (() => {
        const dirDict: {[pieceTypeKey in string]: {[isPromotedKey in  "normal" | "promoted"]: Dir[]}} = {}
        for (const [type, mats] of Object.entries(Piece.MovedirMatrix)) {
            const makeDirPositionList = (mat: integer[][]): Dir[] => {
                if (mat.length === 0) {
                    return []
                }
                const offsetY: integer = Math.floor(mat.length / 2)
                const offsetX: integer = offsetY
                return mat.map((line, y) =>
                    line.map((times, x) => (
                        {dif: {x: x - offsetX, y: y - offsetY}, times}
                    )).filter(dir => dir.times !== 0)
                ).flat()
            }

            dirDict[type as PieceType] = {
                normal: makeDirPositionList(mats[0]),
                promoted: makeDirPositionList(mats[1]),
            }
        }
        return dirDict as  {[pieceTypeKey in PieceType]: {[isPromotedKey in  "normal" | "promoted"]: Dir[]}}
    })()
}
