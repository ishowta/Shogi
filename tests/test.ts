import * as Shogi from '../src/library/shogi/shogi';
import * as Utils from '../src/library/shogi/util';
import * as assert from "power-assert";

const snap = (obj: any, name?: string) => expect(obj).toMatchSnapshot(name)
const readConsoleOutput = (fn: () => void): string => {
    let output = ""
    const storeLog = (inputs: string) => (output += inputs);
    console["log"] = jest.fn(storeLog);
    fn()
    return output
}

/** 何手か適当に進める */
const playComplexMove = (shogi: Shogi.Shogi) => {
    shogi.move(c(7, 7), c(7, 6)) //76歩
    shogi.move(c(3, 3), c(3, 4)) //34歩
    shogi.move(c(8, 8), c(2, 2), true) //22角成,角取り
    shogi.move(c(8, 3), c(8, 4)) //84歩
    const kaku = shogi.hand[Shogi.Player.Black][0]
    shogi.placeHandPiece(kaku, c(6, 2)) //62角打, 王手
}

const c = Shogi.Shogi.c

let shogi: Shogi.Shogi

beforeEach(() => {
    shogi = new Shogi.Shogi()
})

it('インスタンスを作れるか', () => {
    assert(shogi instanceof Shogi.Shogi)
});

it('将棋盤が初期化されているか', () => {
    snap(shogi.board)
})

it('正しく駒を動かせているか', () => {
    playComplexMove(shogi)
    snap(shogi.board)
})

it('将棋盤を表示できるか', () => {
    playComplexMove(shogi)
    snap(readConsoleOutput(() => shogi.printBoard()))
})

it('将棋盤を文字列に変換・逆変換できるか', () => {
    playComplexMove(shogi)
    snap(shogi.toString())
    const newShogi = new Shogi.Shogi()
    newShogi.fromString(shogi.toString())
    assert.deepEqual(newShogi.board, shogi.board)
    assert.deepEqual(newShogi.hand, shogi.hand)
    assert.deepEqual(newShogi.turnPlayer, shogi.turnPlayer)
    // 棋譜は保存されない
    // 🙅 assert(newShogi.score === shogi.score)
})

it('成れるか', () => {
    shogi.move(c(7, 7), c(7, 6)) //76歩
    shogi.move(c(3, 3), c(3, 4)) //34歩
    shogi.move(c(8, 8), c(2, 2), true) //22角成,角取り
    const kaku = shogi.board.at(c(2, 2))
    assert(kaku && kaku.isPromote === true)
})

it('相手の駒は動かせない', () => {
    const res = shogi.move(c(9, 1), c(9, 2))
    assert(res.type === "move_error" && res.reason instanceof Shogi.NoPieceError)
})

it('既に成っているのに成ろうとする', () => {
    shogi.move(c(7, 7), c(7, 6)) //76歩
    shogi.move(c(3, 3), c(3, 4)) //34歩
    shogi.move(c(8, 8), c(2, 2), true) //22角成,角取り
    shogi.move(c(3, 4), c(3, 5)) //34歩
    const res = shogi.move(c(2, 2), c(1, 1), true) //11角成(🙅),香取り
    assert(res.type === "move_error" && res.reason instanceof Shogi.CantPromoteError)
})

it('成ることのできない駒なのに成ろうとする', () => {
    const res = shogi.move(c(5, 9), c(5, 8), true) //58王成(🙅)
    assert(res.type === "move_error" && res.reason instanceof Shogi.CantPromoteError)
})

it('打ち歩詰め', () => {
    shogi.move(c(7, 7), c(7, 6)) //76歩
    shogi.move(c(2, 3), c(2, 4)) //24歩
    shogi.move(c(8, 8), c(3, 3)) //33角,歩取り
    shogi.move(c(4, 1), c(4, 2)) //42金
    const hu = shogi.hand[shogi.turnPlayer][0]
    const res = shogi.placeHandPiece(hu, c(5, 2)) //52歩打,王手(🙅)

    assert(res.type === "put_error" && res.reason instanceof Shogi.StrikingError)
})

it('千日手', () => {
    for (const i of Utils.range(1, 4)) {
        assert(shogi.checkThousandDays() === false)
        shogi.move(c(6, 9), c(6, 8)) //68金
        shogi.move(c(6, 1), c(6, 2)) //62金
        shogi.move(c(6, 8), c(6, 9)) //69金
        shogi.move(c(6, 2), c(6, 1)) //61金
    }
    assert(shogi.checkThousandDays() === true)
})

it('王手放置', () => {
    shogi.move(c(7, 7), c(7, 6)) //76歩
    shogi.move(c(3, 3), c(3, 4)) //34歩
    shogi.move(c(8, 8), c(2, 2)) //22角,角取り
    shogi.move(c(8, 3), c(8, 4)) //84歩
    const kaku = shogi.hand[Shogi.Player.Black][0]
    shogi.placeHandPiece(kaku, c(6, 2)) //62角打, 王手
    const res = shogi.move(c(9, 3), c(9, 4)) //94歩(🙅)

    assert(res.type === "move_error" && res.reason instanceof Shogi.NeglectKingError)
})
