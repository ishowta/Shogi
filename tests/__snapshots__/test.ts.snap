// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`将棋盤が初期化されているか 1`] = `
PieceBoard [
  Array [
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 6,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 0,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 6,
    },
  ],
  Array [
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 1,
    },
    null,
    null,
    null,
    null,
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 2,
    },
    null,
  ],
  Array [
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
  ],
  Array [
    null,
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 2,
    },
    null,
    null,
    null,
    null,
    null,
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 1,
    },
    null,
  ],
  Array [
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 6,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 0,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 6,
    },
  ],
]
`;

exports[`将棋盤を文字列に変換・逆変換できるか 1`] = `"1,,,610-510-410-310-010-310-410-510-610,_-110-_-200-_-_-_-201-_,710-_-710-710-710-710-_-710-710,_-710-_-_-_-_-710-_-_,_-_-_-_-_-_-_-_-_,_-_-700-_-_-_-_-_-_,700-700-_-700-700-700-700-700-700,_-_-_-_-_-_-_-100-_,600-500-400-300-000-300-400-500-600"`;

exports[`将棋盤を表示できるか 1`] = `"(香)(桂)(銀)(金)(玉)(金)(銀)(桂)(香) 　 (飛) 　  角  　  　  　  馬  　 (歩) 　 (歩)(歩)(歩)(歩) 　 (歩)(歩) 　 (歩) 　  　  　  　 (歩) 　  　  　  　  　  　  　  　  　  　  　  　  　  歩  　  　  　  　  　  　  歩  歩  　  歩  歩  歩  歩  歩  歩  　  　  　  　  　  　  　  飛  　  香  桂  銀  金  王  金  銀  桂  香 "`;

exports[`正しく駒を動かせているか 1`] = `
PieceBoard [
  Array [
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 6,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 0,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 6,
    },
  ],
  Array [
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 1,
    },
    null,
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 2,
    },
    null,
    null,
    null,
    Piece {
      "isPromote": true,
      "owner": "black",
      "type": 2,
    },
    null,
  ],
  Array [
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
  ],
  Array [
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    null,
    null,
    null,
    null,
    Piece {
      "isPromote": false,
      "owner": "white",
      "type": 7,
    },
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    null,
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 7,
    },
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 1,
    },
    null,
  ],
  Array [
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 6,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 0,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 3,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 4,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 5,
    },
    Piece {
      "isPromote": false,
      "owner": "black",
      "type": 6,
    },
  ],
]
`;
